x-pack/plugins/fleet/server/services/security/message_signing_service.ts

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { generateKeyPairSync, createSign, randomBytes } from 'crypto';
+
+import type { KibanaRequest } from '@kbn/core-http-server';
+import type { SavedObjectsClientContract } from '@kbn/core-saved-objects-api-server';
+import type { EncryptedSavedObjectsClient } from '@kbn/encrypted-saved-objects-plugin/server';
+import { SECURITY_EXTENSION_ID } from '@kbn/core-saved-objects-server';
+
+import { MESSAGE_SIGNING_KEYS_SAVED_OBJECT_TYPE } from '../../constants';
+import { appContextService } from '../app_context';
+
+interface MessageSigningKeys {
+  private_key: string;
+  public_key: string;
+  passphrase: string;
+}
+
+export interface MessageSigningServiceInterface {
+  getPublicKey(): Promise<string>;
+  sign(message: unknown): Promise<string>;
+  rotateKeys(passphrase: string): Promise<string>;
+}
+
+export class MessageSigningService implements MessageSigningServiceInterface {
+  #esoClient: EncryptedSavedObjectsClient;
+  #_soClient: SavedObjectsClientContract | undefined;
+
+  constructor(esoClient: EncryptedSavedObjectsClient) {
+    this.#esoClient = esoClient;
+  }
+
+  public async getPublicKey(): Promise<string> {
+    const { publicKey } = await this.#getCurrentKeyPair();
+    return publicKey;
+  }
+
+  public async sign(message: unknown): Promise<string> {
+    const signer = createSign('SHA256');
+    signer.update(JSON.stringify(message));
+    signer.end();
+
+    const { privateKey, passphrase } = await this.#getCurrentKeyPair();
+
+    return signer.sign({ key: privateKey, passphrase }, 'hex');
```

Comment:
the output encoding probably going to be 'base64', right? 
initially I assumed base64 in this doc
https://docs.google.com/document/d/1xFdDD8dj1Zw3NwKeYPwyc37VofMHSyM3vA58XR8mimM/edit
which seems to be the format storage for binary data in elasticsearch: 
https://www.elastic.co/guide/en/elasticsearch/reference/current/binary.html